### ISSUE TITLE
Out-of-memory issue

### ISSUE THREAD

**Issue by illuhad** [argonne-lcf/HPC-Patterns#2]
I'm seeing out-of-memory errors with hipSYCL. I think it is related to the max alloc size entering the allocation sizes here:
https://github.com/argonne-lcf/HPC-Patterns/blob/2e8221ba9e62cfd6744b920e08f5b80d8d6dd618/concurency/sycl_con.cpp#L169

It could be that the returned value is not super reliable on hipSYCL, or that multiple allocations are taking place (which the code does not seem to do at first glance?).
What is the best approach to dealing with this? There is the `--globalsize` flag, but this seems to disable the autotuning logic. Is this safe to do? Would it be a better approach to set the default values for these settings in the code to something reasonable?

**Comment by TApplencourt** [comment]
>It could be that the returned value is not super reliable on hipSYCL, or that multiple allocations are taking place (which the code does not seem to do at first glance?).

does hipSYCL implement correctly `sycl::get_pointer_type(ptr, C)`? I was lazy and relied on this query to free my memory.  I can change the code to not rely on this (I will need to do that for the OpenMP porting) 

> What is the best approach to dealing with this? There is the --globalsize flag, but this seems to disable the autotuning logic. Is this safe to do? 

Yes, if you are careful that the "expected speedup" is roughly len(commands) no problem. 

> Would it be a better approach to set the default values for these settings in the code to something reasonable?

The problem is that reason is hard to know... Between Gen9 (who is an integrated GPU so in theory super fast host <-> device datatransfer), A100 (who have a fast PCI), and some general customer GPU (with maybe slow PCI) I give up and relied on the maximum possible memory possible.

I will fix `--globalsize` so you can it can still do auto-tunning.  And add a new `--avoid-autotunning` option. It was a bug :) Thanks for the report!

**Comment by illuhad** [comment]
> does hipSYCL implement correctly sycl::get_pointer_type(ptr, C)? I was lazy and relied on this query to free my memory. I can change the code to not rely on this (I will need to do that for the OpenMP porting)

It should, e.g. for  CUDA:
https://github.com/illuhad/hipSYCL/blob/1046a78777a23ca75c6ea4e92291f1dbe36169ca/src/runtime/cuda/cuda_allocator.cpp#L142
https://github.com/illuhad/hipSYCL/blob/1046a78777a23ca75c6ea4e92291f1dbe36169ca/include/hipSYCL/sycl/usm_query.hpp#L158

I think the problem really is the `info::device::max_mem_alloc_size`:
https://github.com/illuhad/hipSYCL/blob/1046a78777a23ca75c6ea4e92291f1dbe36169ca/src/runtime/cuda/cuda_hardware_manager.cpp#L257
So this seems to return the global memory size at the moment, which is probably incorrect. Presumably this was done since it is the closest thing that `cudaDeviceProp` (which is used to resolve those queries) has to the maximum allocation size:
https://docs.nvidia.com/cuda/cuda-runtime-api/structcudaDeviceProp.html

**Comment by TApplencourt** [comment]
Ok so Intel/{HIP,CUDA} backend is doing :


```
  case PI_DEVICE_INFO_MAX_MEM_ALLOC_SIZE: {
    // Max size of memory object allocation in bytes.
    // The minimum value is max(min(1024 × 1024 ×
    // 1024, 1/4th of CL_DEVICE_GLOBAL_MEM_SIZE),
    // 32 × 1024 × 1024) for devices that are not of type
    // CL_DEVICE_TYPE_HIPSTOM.

    size_t global = 0;
    cl::sycl::detail::pi::assertion(hipDeviceTotalMem(&global, device->get()) ==
                                    hipSuccess);

    auto quarter_global = static_cast<pi_uint32>(global / 4u);

    auto max_alloc = std::max(std::min(1024u * 1024u * 1024u, quarter_global),
                              32u * 1024u * 1024u);

    return getInfo(param_value_size, param_value, param_value_size_ret,
                   pi_uint64{max_alloc});
  }
  ```
  
  (https://cs.github.com/intel/llvm/blob/9b84dd80bbd2068738fe95dfda6a013557eca596/sycl/plugins/hip/pi_hip.cpp?q=max_mem_alloc_size#L1132-L1150)

I don't know why 1/4, maybe some insider information or pure arbitrary choice.

I guess we should do the same things. `min(global_memory/len(command) , max_alloc)` or something like that. Will fix!

**Comment by TApplencourt** [comment]
Fixed by 62b2601

**Comment by illuhad** [comment]
Thanks!

Also thank you for looking into DPC++. I think this is probably some heuristic that just happens to work on common devices. On my GeForce GTX 1080Ti I have done the runs without issue with 4GB memory allocations, which is more than 1/4 device memory... So, DPC++ also does not seem to have the "true" maximum allocation size..
